### PR TITLE
PTF: add basic test infra and more test cases

### PR DESCRIPTION
**Pull request by vmytnykx**
This patch-set introduce PTF test cases and basic infrastructure (similar to SONiC PTF) to run PTF DASH tests. For now the test cases supports minimal set of use cases, but it is the base for future test case implementations that will cover whole overlay PTF test scenario (in progress).

**Main changes:**

- Add basic PTF infra (copied from SONiC PTF) for running PTF, allows run test cases w/o SAI repo, only PTF is required. DASH repo becomes main repo for test development.
- Add Common VNet API for all tests (to void duplicate work, can be split from this file later)
- Add the following test cases:
  - Inbound VNI/ENI
  - Inbound Routing PA validate
  - Inbound Routing VNI match
  - Inbound Routing PA SRC IP
  - Outbound Routing Vnet direct
  - Outbound Routing direct
  - CT: just placeholder, can be used for debugging only
  - L3 Route basic: just to make sure traffic is routed
- Run only bmv2 ready test case on CI because of missing functionality in bmv2. Once more bmv2 functionality becomes available, test_saithrift_vnet.py can be merged with this one
- test_saithrift_vnet: removed unused libs

**Future plan:**
- Topology/setup diagram
- Test plan (in progress)
- Check/validate BMV2 support
- More CI integration
- More updates and fixes
- Update TCs to SAI PTF repo (once stable)

### CODE REVIEW

**Review by chrispsommers — COMMENTED** [review]
Hi @vmytnykx I am not sure why copying `platform_helper/` is necessary since SAI is already a submodule in the dash project. You have access to everything under SAI/PTF already.  In addition, the saithrift-client docker image has a copy of SAI/ptf directory embedded in it, see link below. Is this mainly for convenience so you have these utils in your host workspace in a "nearby" directory which makes importing easier? The philosophy was to put all tools in the container and not rely on host workspace (but that's subjective). My main concern with copying is that it doesn't automatically keep up with changes in the original repo (SAI) and can create a maintenance issue in the future.

Nevertheless, I don't see anything "wrong" with this change, it just might not be needed.

https://github.com/Azure/DASH/blob/d003f4e95fcf6438224782d3f1ee9e3aa53d17d5/dash-pipeline/dockerfiles/Dockerfile.saithrift-client#L16

**Comment by KrisNey-MSFT** [comment]
@prsunny  would you please take a look comment?  Once we have stabilized, we can push more files.

**Comment by chrispsommers** [comment]
@mgheorghe Please note your tests were moved into new `scale/` directory. Are you OK with this?

File: test/test-cases/README.md [hunk 1]
```diff
@@ -3,6 +3,20 @@ This contains a hierarchical set of directories containing DASH test cases organ
 
 # Contents
 
-| Folder                                         | Description                                               |
-| ---------------------------------------------- | --------------------------------------------------------- |
-| [vnet2vnet](vnet2vnet/README.md)                  | DASH vnet2vnet Tests 
+| Folder | Description |
+| --- | --- |
+| functional | Tests to verify essential functionality using low-rate traffic (SAI PTF).
+| scale | Tests with hight-rate traffic and complex configuration to verify scaling real-world scenarios.
```
Comment (chrispsommers) [inline test/test-cases/README.md:9]: hight - typo

File: test/test-cases/README.md [hunk 2]
```diff
@@ -3,6 +3,20 @@ This contains a hierarchical set of directories containing DASH test cases organ
...
+
+## functional
+
+| Folder/File | Description |
+| --- | --- |
+| saidashvnet.py | VNET-to-VNET test cases
```
Comment (chrispsommers) [inline test/test-cases/README.md:15]: Need to add `saidashvnet` to .wordlist.txt to pass spellcheck.

**Review by chrispsommers — CHANGES_REQUESTED** [review]
See spellcheck issues.

File: test/test-cases/README.md (hunk 2 above)
Comment (chrispsommers) [reply test/test-cases/README.md:15]: There might be some more customization available, see https://github.com/rojopolis/spellcheck-github-actions

Comment (aputriax) [reply test/test-cases/README.md:15]: I used backticks for the filenames. It works.

**Review by chrispsommers — APPROVED** [review]

**Comment by chrispsommers** [comment]
Hi @vmytnykx just checking with you, is this read to merge in your opinion? LGTM.

**Comment by chrispsommers** [comment]
@aputriax if we merge this now, will it impact sai-challenger pending merges in any significant way?

**Comment by mgheorghe** [comment]
the images folder with diagrams disappeared, also the links to pictures were not updated in the MD files, this PR needs few fixes

**Comment by chrispsommers** [comment]
This was already merged, can @mgheorghe  and @aputriax please figure out if a remedial PR is required? Thanks.

**Comment by aputriax** [comment]
@mgheorghe, the images folder stays as is on the same place, only the link has been broken due to renaming. I've created PR #262 with fixes. I also added a missed link to one README.md and fixed spellchecker. Please verify it. Thanks!

File: test/test-cases/functional/saidashvnet.py [hunk 3]
```diff
@@ -0,0 +1,763 @@
...
+        sai_thrift_remove_outbound_routing_entry(self.client, entry)
+
+    def outbound_ca_to_pa_create(self, dst_vnet_id, dip, underlay_dip):
+        """
+        Create outband CA PA mapping
+        """
+
+        ca_to_pa_entry = sai_thrift_outbound_ca_to_pa_entry_t(switch_id=self.switch_id,
+                                                              dst_vnet_id=dst_vnet_id, dip=sai_ipaddress(dip))
+        sai_thrift_create_outbound_ca_to_pa_entry(self.client, ca_to_pa_entry,
```
Comment (vincent-xs) [inline test/test-cases/functional/saidashvnet.py:238]: @vmytnykx 
as per p4 outbound code, the overlay dmac is always provided as part of the ca to pa mapping data.

 action set_tunnel_mapping(IPv4Address underlay_dip,
                              EthernetAddress overlay_dmac,
                              bit<1> use_dst_vnet_vni) {
        if (use_dst_vnet_vni == 1)
            meta.vnet_id = meta.dst_vnet_id;
        meta.encap_data.overlay_dmac = overlay_dmac;
        meta.encap_data.underlay_dip = underlay_dip;
    }

any reason the dmac is not provided in those testcases?

File: test/test-cases/functional/saidashvnet.py [hunk 4]
```diff
@@ -0,0 +1,763 @@
...
+        out_pkt = simple_udp_packet(eth_src="00:00:00:01:03:14",
+                                    eth_dst="20:30:40:50:60:70",
+                                    ip_dst="10.10.2.2",
+                                    ip_src="10.10.20.20",
+                                    ip_ttl=64)
+        exp_pkt = simple_udp_packet(eth_src=self.RIF_SRC_MAC,
+                                    eth_dst=self.NEIGH_DMAC,
+                                    ip_dst="10.10.2.2",
+                                    ip_src="10.10.20.20",
+                                    ip_ttl=64)
```
Comment (vincent-xs) [inline test/test-cases/functional/saidashvnet.py:759]: since the packet is routed, shouldn't the TTL be decremented?
